From 3844d9f336c948ec4f2f66fb8c4dc256618d3c74 Mon Sep 17 00:00:00 2001
From: Rahul Lakkireddy <rahul.lakkireddy@chelsio.com>
Date: Fri, 29 Sep 2017 12:09:09 +0530
Subject: [PATCH 33/33] iw_cxgb4: remove the stid on listen create failure

commit 8b1bbf36b7452c4acb20e91948eaa5e225ea6978
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Tue Sep 5 11:52:34 2017 -0700

    iw_cxgb4: remove the stid on listen create failure

    If a listen create fails, then the server tid (stid) is incorrectly left
    in the stid idr table, which can cause a touch-after-free if the stid
    is looked up and the already freed endpoint is touched.  So make sure
    and remove it in the error path.

    Cc: stable@vger.kernel.org
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>
    Signed-off-by: Doug Ledford <dledford@redhat.com>
---
 drivers/infiniband/hw/cxgb4/cm.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/infiniband/hw/cxgb4/cm.c b/drivers/infiniband/hw/cxgb4/cm.c
index 5c10a034..4691586b 100644
--- a/drivers/infiniband/hw/cxgb4/cm.c
+++ b/drivers/infiniband/hw/cxgb4/cm.c
@@ -3502,7 +3502,7 @@ int c4iw_create_listen(struct iw_cm_id *cm_id, int backlog)
 		cm_id->provider_data = ep;
 		goto out;
 	}
-
+	remove_handle(ep->com.dev, &ep->com.dev->stid_idr, ep->stid);
 	cxgb4_free_stid(ep->com.dev->rdev.lldi.tids, ep->stid,
 			ep->com.local_addr.ss_family);
 fail2:
-- 
2.14.1

